build_tools/testing/run_python_notebook.sh

```diff
@@ -0,0 +1,52 @@
+#!/bin/bash
+# Copyright 2021 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+# Run a Python (Colab/Jupyter) notebook in an isolated virtual environment.
+# Fails if the notebook has errors.
+#
+# Usage: run_python_notebook.sh /path/to/notebook.ipynb
+#
+# This is intended for use on continuous integration servers within a Docker
+# container and as a reference for users, but it can also be run locally.
+
+set -x
+set -e
+
+# Run under a virtual environment to isolate Python packages.
+#
+# Note: --without-pip and --system-site-packages are workarounds for Docker
+#  * https://stackoverflow.com/a/63805343
+#  * https://askubuntu.com/a/897004
+#
+# Note: --clear deletes existing venv contents, which helps make these tests
+# more hermetic, at the cost of time spent performing redundant installs.
+# TODO(scotttodd): evaluate if we can do without --clear
+python3 -m venv .venv --without-pip --system-site-packages --clear
+source .venv/bin/activate
+trap deactivate EXIT
+
+# Install general Jupyter notebook requirements.
+python3 -m pip install --quiet jupyter_core nbconvert ipykernel
+
+# Install common notebook requirements.
+# TODO(scotttodd): refactor so not all deps are always installed
+python3 -m pip install --quiet \
+  numpy \
+  matplotlib \
+  tensorflow \
+  tensorflow_hub \
+  bottleneck
```

Comment:
The Colab notebooks `pip install` any packages that are not provided as part of Colab's default hosted runtimes, e.g. https://colab.research.google.com/github/google/iree/blob/main/colab/low_level_invoke_function.ipynb#scrollTo=Go2Nw7BgIHYU&line=1&uniqifier=1 . That's mostly what the `venv` helps with (so we test that notebooks are installing what they need, which could be pinned to specific versions or pulling from the latest snapshots).

The common requirements listed here _could_ be installed system-wide. I can fiddle with the `--without-pip --system-site-packages` flags a bit more to see if that will work. I'd actually prefer that, since installing TF (and to a lesser extent the others) can take a few minutes.

Reply:
Yeah installing them in the docker image makes sense to me. You might then want to base yourself off the image we use for our integrations tests. Note that it uses tf-nightly at a pinned version

Reply:
Reworked to install in the docker image then attempt to reuse in the venv. Seems to work, but the whole venv setup within Docker is pretty brittle. Hopefully the comments + stackoverflow links are enough for future maintainers.